firedrake/plot.py

```diff
@@ -292,6 +294,321 @@ def quiver(function, **kwargs):
+
+        if index != 0:
+            self._enter_distance_to_chunk(chunk[:index])
+            chunks.append(chunk[:index])
+
+        backward_streamline = np.concatenate(chunks, axis=0)
+
+        streamline = np.vstack((backward_streamline[:0:-1], forward_streamline))
+        self.streamlines.append(streamline)
+
+
+def streamplot(function, resolution=None, min_length=None, max_time=None,
+               start_width=0.5, end_width=1.5, tolerance=3e-3, loc_tolerance=1e-10,
+               seed=None, **kwargs):
+    r"""Create a streamline plot of a vector field
+
+    Similar to matplotlib :func:`streamplot <matplotlib.pyplot.streamplot>`
+
+    :arg function: the Firedrake :class:`~.Function` to plot
+    :arg resolution: minimum spacing between streamlines (defaults to domain size / 20)
+    :arg min_length: minimum length of a streamline (defaults to 4x resolution)
+    :arg max_time: maximum time to integrate a streamline
+    :arg start_width: line width at beginning of streamline
+    :arg end_width: line width at end of streamline, to convey direction
+    :arg tolerance: dimensionless tolerance for adaptive ODE integration
+    :arg loc_tolerance: point location tolerance for :meth:`~firedrake.functions.Function.at`
+    :kwarg kwargs: same as for matplotlib :class:`~matplotlib.collections.LineCollection`
+    """
+    if function.ufl_shape != (2,):
+        raise ValueError("Streamplot only defined for 2D vector fields!")
+
+    axes = kwargs.pop("axes", None)
+    if axes is None:
+        figure = plt.figure()
+        axes = figure.add_subplot(111)
+
+    mesh = function.ufl_domain()
+    if resolution is None:
+        coords = mesh.coordinates.dat.data_ro
+        resolution = (coords.max(axis=0) - coords.min(axis=0)).max() / 20
+
+    if min_length is None:
+        min_length = 4 * resolution
+
+    if max_time is None:
+        area = assemble(Constant(1) * dx(mesh))
+        average_speed = np.sqrt(assemble(inner(function, function) * dx) / area)
+        max_time = 50 * min_length / average_speed
+
+    streamplotter = Streamplotter(function, resolution, min_length, max_time,
+                                  tolerance, loc_tolerance)
+
+    # TODO: better way of seeding start points
+    start_points = []
+    shape = streamplotter._grid.shape
+    for i in range(shape[0] - 1):
+        for j in range(shape[1] - 1):
+            start_points.append(streamplotter._grid_point((i, j)))
+
+    # Randomly shuffle the start points
+    import randomgen
+    generator = randomgen.MT19937(seed).generator
+    start_points = generator.permutation(np.array(start_points))
+
+    for x in start_points:
+        streamplotter.add_streamline(x)
+
+    # Colors are determined by the speed, thicknesses by arc length
+    speeds = []
+    widths = []
+    for streamline in streamplotter.streamlines:
+        velocity = np.array(function.at(streamline, tolerance=loc_tolerance))
```

Comment:
The fact that function.at doesn't return a numpy array seems to be annoyance. Should we fix that (maybe separately). WDYT?

Reply:
I think returning a numpy array makes more sense, especially given that you can preallocate it easily, but that seems like something for another PR.

Reply:
OK.